feat: merge operation

Blajda opened the pull request:
# Description
Implement the Merge operation using Datafusion.

Currently the implementation rewrites the entire DeltaTable limiting the files that are rewritten will be performed in future work. 

# Related Issue(s)
- progresses #850


# Documentation

<!---
Share links to useful documentation
--->

ion-elgreco commented:
@Blajda @wjones127 does that mean the full table will be loaded in memory first so it can rewrite the whole table?

Blajda commented:
>  does that mean the full table will be loaded in memory first so it can rewrite the whole table?

Yes since this is using a nested loop join both the source and target need to be fully loaded into memory. There would need to be follow up improvements to perform pruning and to use a Hash Join / Merge Join.

ion-elgreco commented:
> >  does that mean the full table will be loaded in memory first so it can rewrite the whole table?
> 
> Yes since this is using a nested loop join both the source and target need to be fully loaded into memory. There would need to be follow up improvements to perform pruning and to use a Hash Join / Merge Join.

Good to know, still useful for datasets that fit in memory then. Hopefully it can be `merged` soon ;) Then I could try taking a go on the Python bindings and start using it as :)

wjones127 commented on rust/src/operations/merge.rs:
```diff
@@ -0,0 +1,1614 @@
+    ///     .merge(source, col("id").eq(col("source.id")))
+    ///     .with_source_alias("source")
+    ///     .when_matched_delete(|delete| {
+    ///         update
```
This meant to be `delete`?

wjones127 commented on rust/src/operations/merge.rs:
```diff
@@ -0,0 +1,1614 @@
+    ///         update
+    ///             .predicate(col("source.delete"))
+    ///     })?
+    ///     .when_matched_update(|update| {
+    ///         update
+    ///             .update("value", col("source.value"))
+    ///             .update("modified", col("source.modified"))
+    ///     })?
```
should this be delete or is this meant to show how to mix delete and update?

wjones127 commented on rust/src/operations/merge.rs:
````diff
@@ -0,0 +1,1614 @@
+    /// Only the first clause that results in an satisfy predicate is executed.
+    /// Ther order of not match clauses matter.
+    ///
+    /// #Example
+    /// ```rust ignore
````
I don't think the example uses this method. Does it?

wjones127 commented on rust/src/operations/merge.rs:
```diff
@@ -0,0 +1,1614 @@
+
+    // TODO: Currently a NestedLoopJoin is used but we should target to support SortMergeJoin
+    // This would require rewriting the join predicate to only contain equality between left and right columns and pushing some filters down
+    // Ideally it would be nice if the optimizer / planner can pick the best join so maybe explore rewriting the entire operation using logical plans.
```
Agreed. It would be nice if we could write this instead with the DataFrame API or as a logical plan, rather than having to write it as a physical one.

wjones127 commented on rust/src/operations/merge.rs:
```diff
@@ -0,0 +1,1614 @@
+//! and specify additional predicates for finer control. The order of operations
+//! specified matter.  See [`MergeBuilder`] for more information
+//!
+//! *WARNING* The current implementation rewrites the entire delta table so only
+//! use on small to medium sized tables. Enhancements tracked at #850
```
Also worth noting that it currently loads the entire thing into memory and uses only 1 thread, IIUC.

wjones127 commented on rust/src/operations/merge.rs:
```diff
@@ -0,0 +1,1614 @@
+            parameters["notMatchedPredicates"],
+            json!(r#"[{"actionType":"insert"}]"#)
+        );
+        // Todo: Expected this predicate to actually be 'value = 1'. Predicate should contain a valid sql expression
+        assert_eq!(
+            parameters["notMatchedBySourcePredicates"],
+            json!(r#"[{"actionType":"update","predicate":"value = Int32(1)"}]"#)
+        );
```
Should this be fixed before merging?

Blajda replied:
Yeah I think we can hold on merging this until it's resolved. I raised an ask on the DF tracker here: https://github.com/apache/arrow-datafusion/issues/7165
This same issue exits for delete and update too so they'll need to be fixed too.

wjones127 commented on rust/src/operations/update.rs:
```diff
@@ -291,7 +288,22 @@ async fn execute(
         Arc::new(ProjectionExec::try_new(expressions, parquet_scan)?);
 
-    let count_plan = Arc::new(UpdateCountExec::new(projection_predicate.clone()));
+    let count_plan = Arc::new(MetricObserverExec::new(
```
Nice job abstraction this!

Blajda replied:
Thanks :)

wjones127 reviewed (changes requested):
Sorry it took me so long to get to this.

This is a big operation, thanks for starting this.

I think since we are using the inner loop join, I don't think we can expect great performance out of this except in very small tables. But it will take some work to move towards logical plans that can choose more efficient joins. So for this PR, I think it's best we mostly focus on correctness. The main thing I see is that the record SQL statement isn't quite correct.

Blajda commented:
No worries about the delays I understand efforts on this project are mostly volunteer and I appreciate your time :)

I've updated the docs with the correct methods and refined description for each method.

What do you think about putting this operation behind a feature flag? Anyone using it would opt in having understood the current issues and it allow smaller PRs to be submitted for the enhancements.

For next steps, I have another branch where I introduce a `DeltaScanBuilder` and updated `DeltaScan` to include a  optional column that maintains which file a particular record came from. The implementation is similar to the `FindFiles` and would be used here to determine which files actually have modifications.

wjones127 commented:
> What do you think about putting this operation behind a feature flag?

I don't feature flag is necessary, but we could label it "experimental" maybe. I think as long as the docs give up-to-date notes on limitations this should be fine.

Blajda commented:
Thanks @wjones127  for taking the time to review this. I'll break future enhancements into smaller chunks to make it easier.

roeap commented:
Hoping that merging main will resolve some of the failures, which seem to be related to recently added deletion vector fields...

Blajda commented:
Updated code to add new dv field. hopefully it passes all test now.

roeap commented:
There is one more failure - probably related to a mypy update o.a. - we could just `# type: ignore` that line for now.

Other then that we are good to go :)

Blajda commented:
:confused: Seems to fail with the ignore and without. check-python is passes locally for me

roeap commented:
yea, this is a bit annoying :( - unfortunately I do not have permissions to disable or bypass required checks ... 

Should we try once more without? Local checks are passing for me as well without the comment.

Alternatively we could try `# noqa`, which is quite common, but not sure if mypy honors it.

roeap commented:
or maybe even both, not sure how make works. i.e. would it fail early if now the ruff check is complaining? otherwise `# noqa` would silence ruff ...

Blajda commented:
@roeap @wjones127 Please provide approval on the workflows. The lint change should hopefully allow everything to pass now.

ion-elgreco commented:
@roeap @wjones127 can you approve the workflow, so this can get merged in the next release?

Blajda commented:
@wjones127 Please approve the workflow. Merged with the latest changes in main.

Blajda commented:
@wjones127 Thanks for reviewing and merging.
